File: backend/app/schemas/analytics.py
from datetime import datetime, date
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field, validator
from app.models.models import RegionEnum


class AnalyticsRequest(BaseModel):
    facebook_page_id: Optional[int] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    metrics: Optional[List[str]] = Field(None, max_items=20)

    @validator('end_date')
    def validate_date_range(cls, v, values):
        if 'start_date' in values and values['start_date'] and v:
            if v <= values['start_date']:
                raise ValueError('End date must be after start date')
            if (v - values['start_date']).days > 90:
                raise ValueError('Date range cannot exceed 90 days')
        return v


class PostAnalyticsResponse(BaseModel):
    id: int
    facebook_page_id: int
    scheduled_post_id: Optional[int]
    impressions: int
    reach: int
    engaged_users: int
    clicks: int
    likes: int
    comments: int
    shares: int
    reactions_love: int
    reactions_wow: int
    reactions_haha: int
    reactions_sad: int
    reactions_angry: int
    engagement_rate: float
    click_through_rate: float
    cost_per_engagement: Optional[float]
    performance_score: float
    relative_performance: Optional[str]
    last_updated: datetime
    created_at: datetime

    class Config:
        from_attributes = True


class PageAnalyticsSummary(BaseModel):
    facebook_page_id: int
    date_range: Dict[str, date]
    total_posts: int
    total_impressions: int
    total_reach: int
    total_engaged_users: int
    total_clicks: int
    avg_engagement_rate: float
    avg_click_through_rate: float
    best_performing_post: Optional[Dict[str, Any]]
    worst_performing_post: Optional[Dict[str, Any]]
    posting_frequency: float
    growth_metrics: Dict[str, float]


class PerformanceComparison(BaseModel):
    current_period: PageAnalyticsSummary
    previous_period: PageAnalyticsSummary
    improvements: Dict[str, float]
    declines: Dict[str, float]
    recommendations: List[str]


class OptimizationInsightResponse(BaseModel):
    id: int
    facebook_page_id: int
    insight_type: str
    insight_data: Dict[str, Any]
    confidence_score: float
    recommendation: str
    expected_improvement: Optional[float]
    is_implemented: bool
    implementation_date: Optional[datetime]
    actual_improvement: Optional[float]
    created_at: datetime
    expires_at: Optional[datetime]

    class Config:
        from_attributes = True


class RegionalPerformanceComparison(BaseModel):
    us_performance: Optional[PageAnalyticsSummary]
    uk_performance: Optional[PageAnalyticsSummary]
    regional_insights: List[str]
    cross_regional_recommendations: List[str]


class AnalyticsDashboard(BaseModel):
    summary: PageAnalyticsSummary
    recent_posts: List[PostAnalyticsResponse]
    optimization_insights: List[OptimizationInsightResponse]
    performance_trends: Dict[str, List[float]]
    upcoming_optimizations: List[Dict[str, Any]]


class EngagementTimeline(BaseModel):
    facebook_page_id: int
    timeline_data: List[Dict[str, Any]]  # Date and engagement metrics
    peak_hours: List[int]
    best_days: List[str]
    seasonal_patterns: Dict[str, Any]


class ContentPerformanceAnalysis(BaseModel):
    content_type_performance: Dict[str, float]
    optimal_caption_length: int
    best_hashtag_count: int
    sentiment_impact: Dict[str, float]
    image_vs_text_performance: Dict[str, float]
    regional_preferences: Dict[str, Any]

File: frontend/src/services/analytics.ts

import { apiCall, type ApiResponse } from './api'
import type { 
  AnalyticsDashboard,
  PageAnalyticsSummary,
  PostAnalytics,
  OptimizationInsight
} from '@/types'

interface PerformanceComparison {
  current_period: PageAnalyticsSummary
  previous_period: PageAnalyticsSummary
  improvements: Record<string, number>
  declines: Record<string, number>
  recommendations: string[]
}

interface RegionalPerformanceComparison {
  us_performance?: PageAnalyticsSummary
  uk_performance?: PageAnalyticsSummary
  regional_insights: string[]
  cross_regional_recommendations: string[]
}

interface EngagementTimeline {
  facebook_page_id: number
  timeline_data: Array<{
    date: string
    avg_engagement: number
    total_reactions: number
    post_count: number
  }>
  peak_hours: number[]
  best_days: string[]
  seasonal_patterns: Record<string, any>
}

interface ContentPerformanceAnalysis {
  content_type_performance: Record<string, number>
  optimal_caption_length: number
  best_hashtag_count: number
  sentiment_impact: Record<string, number>
  image_vs_text_performance: Record<string, number>
  regional_preferences: Record<string, any>
}

class AnalyticsService {
  async getDashboard(
    pageId?: number, 
    days: number = 30
  ): Promise<ApiResponse<AnalyticsDashboard>> {
    return apiCall<AnalyticsDashboard>({
      method: 'GET',
      url: '/api/v1/analytics/dashboard',
      params: { page_id: pageId, days },
    })
  }

  async getPageSummary(
    pageId: number,
    startDate?: string,
    endDate?: string
  ): Promise<ApiResponse<PageAnalyticsSummary>> {
    return apiCall<PageAnalyticsSummary>({
      method: 'GET',
      url: `/api/v1/analytics/pages/${pageId}/summary`,
      params: { start_date: startDate, end_date: endDate },
    })
  }

  async getPostAnalytics(postId: number): Promise<ApiResponse<PostAnalytics>> {
    return apiCall<PostAnalytics>({
      method: 'GET',
      url: `/api/v1/analytics/posts/${postId}`,
    })
  }

  async collectAnalytics(pageId: number): Promise<ApiResponse<{ message: string }>> {
    return apiCall<{ message: string }>({
      method: 'POST',
      url: `/api/v1/analytics/collect/${pageId}`,
    })
  }

  async comparePerformance(
    pageId: number,
    daysCurrent: number = 30,
    daysPrevious: number = 30
  ): Promise<ApiResponse<PerformanceComparison>> {
    return apiCall<PerformanceComparison>({
      method: 'GET',
      url: '/api/v1/analytics/compare',
      params: { 
        page_id: pageId, 
        days_current: daysCurrent, 
        days_previous: daysPrevious 
      },
    })
  }

  async getRegionalComparison(
    days: number = 30
  ): Promise<ApiResponse<RegionalPerformanceComparison>> {
    return apiCall<RegionalPerformanceComparison>({
      method: 'GET',
      url: '/api/v1/analytics/regional-comparison',
      params: { days },
    })
  }

  async getEngagementTimeline(
    pageId: number,
    days: number = 30
  ): Promise<ApiResponse<EngagementTimeline>> {
    return apiCall<EngagementTimeline>({
      method: 'GET',
      url: `/api/v1/analytics/engagement-timeline/${pageId}`,
      params: { days },
    })
  }

  async getContentAnalysis(
    pageId: number,
    days: number = 90
  ): Promise<ApiResponse<ContentPerformanceAnalysis>> {
    return apiCall<ContentPerformanceAnalysis>({
      method: 'GET',
      url: `/api/v1/analytics/content-analysis/${pageId}`,
      params: { days },
    })
  }

  async getOptimizationInsights(
    pageId?: number
  ): Promise<ApiResponse<OptimizationInsight[]>> {
    return apiCall<OptimizationInsight[]>({
      method: 'GET',
      url: '/api/v1/analytics/optimization-insights',
      params: { page_id: pageId },
    })
  }
}

export const analyticsService = new AnalyticsService()
